overlord/devicestate/devicestate.go

```diff
@@ -341,49 +341,99 @@ func extractDownloadInstallEdgesFromTs(ts *state.TaskSet) (firstDl, lastDl, firs
+			return true, new.KernelTrack()
+		}
+		trackChange := current.KernelTrack() != new.KernelTrack()
+		if trackChange {
+			return true, new.KernelTrack()
+		}
+	} else {
+		if current.Kernel() != new.Kernel() {
+			return true, new.KernelSnap().DefaultChannel
+		}
+		defaultChannelChange := current.KernelSnap().DefaultChannel != new.KernelSnap().DefaultChannel
+		if defaultChannelChange {
+			return true, new.KernelSnap().DefaultChannel
+		}
+	}
+	return false, ""
+}
+
+func gadgetChange(current, new *asserts.Model) (changed bool, newGadgetChange string) {
+	if new.Grade() == asserts.ModelGradeUnset {
+		if current.Gadget() != new.Gadget() {
+			return true, new.GadgetTrack()
+		}
+		trackChange := current.GadgetTrack() != new.GadgetTrack()
+		if trackChange {
+			return true, new.GadgetTrack()
+		}
+	} else {
+		if current.Gadget() != new.Gadget() {
+			return true, new.GadgetSnap().DefaultChannel
+		}
+		defaultChannelChange := current.GadgetSnap().DefaultChannel != new.GadgetSnap().DefaultChannel
+		if defaultChannelChange {
+			return true, new.GadgetSnap().DefaultChannel
+		}
+	}
+	return false, ""
+}
+
 func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Model, deviceCtx snapstate.DeviceContext, fromChange string) ([]*state.TaskSet, error) {
 	userID := 0
 	var tss []*state.TaskSet
 
 	// kernel
-	if current.Kernel() == new.Kernel() && current.KernelTrack() != new.KernelTrack() {
-		ts, err := snapstateUpdateWithDeviceContext(st, new.Kernel(), &snapstate.RevisionOptions{Channel: new.KernelTrack()}, userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
-		if err != nil {
-			return nil, err
-		}
-		tss = append(tss, ts)
-	}
-
-	var ts *state.TaskSet
-	if current.Kernel() != new.Kernel() {
-		needsInstall, err := notInstalled(st, new.Kernel())
-		if err != nil {
-			return nil, err
-		}
-		if needsInstall {
-			ts, err = snapstateInstallWithDeviceContext(ctx, st, new.Kernel(), &snapstate.RevisionOptions{Channel: new.KernelTrack()}, userID, snapstate.Flags{}, deviceCtx, fromChange)
+	kernelChanged, newKernelChannel := kernelChange(current, new)
+	if kernelChanged {
+		if current.Kernel() == new.Kernel() {
+			ts, err := snapstateUpdateWithDeviceContext(st, new.Kernel(),
+				&snapstate.RevisionOptions{Channel: newKernelChannel},
+				userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
+			if err != nil {
+				return nil, err
+			}
+			tss = append(tss, ts)
 		} else {
-			ts, err = snapstate.LinkNewBaseOrKernel(st, new.Base())
-		}
-		if err != nil {
-			return nil, err
+			needsInstall, err := notInstalled(st, new.Kernel())
+			if err != nil {
+				return nil, err
+			}
+			var ts *state.TaskSet
+			if needsInstall {
+				ts, err = snapstateInstallWithDeviceContext(ctx, st, new.Kernel(), &snapstate.RevisionOptions{Channel: newKernelChannel}, userID, snapstate.Flags{}, deviceCtx, fromChange)
+			} else {
+				ts, err = snapstate.LinkNewBaseOrKernel(st, new.Base())
```

Comment:
this seems to have been untested, new.Base should be new.Kernel?  also for UC20 we still need to do an update if the channel has changed, no?